src/classes/logger.py

```diff
@@ -40,65 +40,96 @@
+        self.predicate = predicate
+
+    def filter(self, record):
+        source = getattr(record, "source", "")
+        return self.predicate(source)
+
 
 # Set up log formatters
 template = '%(levelname)s %(module)s: %(message)s'
 console_formatter = logging.Formatter(template)
 file_formatter = logging.Formatter('%(asctime)s ' + template, datefmt='%H:%M:%S')
 
-# Add normal stderr stream handler
-sh = logging.StreamHandler()
-sh.setFormatter(console_formatter)
-sh.setLevel(info.LOG_LEVEL_CONSOLE)
-log.addHandler(sh)
-handlers['stream'] = sh
+# Configure root logger for minimal logging
+logging.basicConfig(level=logging.ERROR)
+root_log = logging.getLogger()
+
+# Set up our top-level logging context
+log = root_log.getChild('OpenShot')
+log.setLevel(info.LOG_LEVEL_FILE)
+# Don't pass messages on to root logger
+log.propagate = False
 
-# Add rotating file handler
+#
+# Create rotating file handler
+#
 fh = logging.handlers.RotatingFileHandler(
-    os.path.join(info.USER_PATH, 'openshot-qt.log'), encoding="utf-8", maxBytes=25*1024*1024, backupCount=3)
-fh.setFormatter(file_formatter)
+         os.path.join(info.USER_PATH, 'openshot-qt.log'),
+         encoding="utf-8",
+         maxBytes=25*1024*1024, backupCount=3)
 fh.setLevel(info.LOG_LEVEL_FILE)
+fh.setFormatter(file_formatter)
+
 log.addHandler(fh)
-handlers['file'] = fh
+
+#
+# Create typical stream handler which logs to stderr
+#
+sh = logging.StreamHandler(sys.stderr)
+sh.setLevel(info.LOG_LEVEL_CONSOLE)
+sh.setFormatter(console_formatter)
+
+# Filter out redirected output on console, to avoid duplicates
+#
+# The lambda is used to test the "source" property of the
+# logging record, if present.
+# {"source": "stream"} will be attached to any records generated
+# by the redirection, thanks to LoggingAdapter)
+filt = StreamFilter(lambda x: x != "stream")
+sh.addFilter(filt)
+
+log.addHandler(sh)
 
 
 def reroute_output():
     """Route stdout and stderr to logger (custom handler)"""
-    if not getattr(sys, 'frozen', False):
-        # Hang on to the original objects
-        streams.update({
-            'stderr': sys.stderr,
-            'stdout': sys.stdout,
-            })
-        # Re-route output streams
-        handlers['stdout'] = StreamToLogger(log, logging.INFO)
-        sys.stdout = handlers['stdout']
-        handlers['stderr'] = StreamToLogger(log, logging.ERROR)
-        sys.stderr = handlers['stderr']
+    if (getattr(sys, 'frozen', False)
+       or sys.stdout != sys.__stdout__):
+        return
+    sys.stdout = StreamToLogger(sys.stdout, logging.INFO)
+    sys.stderr = StreamToLogger(sys.stderr, logging.WARNING)
 
 
 def set_level_file(level=logging.INFO):
-    handlers['file'].setLevel(level)
+    """Adjust the minimum log level written to our logfile"""
+    fh.setLevel(level)
 
 
 def set_level_console(level=logging.INFO):
-    handlers['stream'].setLevel(level)
-
+    """Adjust the minimum log level for output to the terminal"""
+    sh.setLevel(level)
```

Comment:
Nitpick: is there a blank line at the end of this file? Or is the PR just not showing it... :thinking:

Reply:
The file ENDS with a newline (my editor makes sure of that automatically), but I'm not sure there's a _blank_ line. I can check.

Reply:
So, turns out, there was a surprising amount of nuance to that question.

Like I said, the file **ended** on a blank line, my editor is set to make sure of that. (Actually, thanks to the `.editorconfig` file at the root of the repo, so is any editor that respects those files.) That means that the file is protected against concatenation damage and other issues, so you could do this:
```sh
$ echo "=======" >> logger.py; tail -3 logger.py
    """Adjust the minimum log level for output to the terminal"""
    sh.setLevel(level)
=======
```
You'd **THINK** it would also mean that an editor would let you position your cursor on line 136, since the last line of the file (line 135) ends with a newline. But surprisingly enough, the first two editors I checked (`vim` and `gedit`) didn't; they eat the final newline, so in those editors the last line of the file indeed appeared to be line 135.

So, I added a _second_ newline to the end of the file, ensuring that editors would show a blank line at the end.

...Naturally, my own editor (Github Atom) **doesn't** eat the final newline, so in Atom it now appears as though there are _two_ blank lines at the end. :rofl: :

![image](https://user-images.githubusercontent.com/538020/120887347-3ba51780-c5c0-11eb-9138-cc74fd63417e.png)